### ISSUE TITLE
Cert rejected in-app with Roblox 

### ISSUE THREAD

**Issue by yaeleiger** [httptoolkit/frida-android-unpinning#50]
After using [this](https://httptoolkit.com/blog/frida-certificate-pinning/) great tutorial, I can confirm twitter's certificate un-pinning and can see traffic from twitter. Unfortunately, when I try to do the same with Roblox, I still can't get passed their cert pining. In my mitmdump I see: "Client TLS handshake failed. The client does not trust the proxy's certificate for *.roblox.com"

And in app, I see the roblox alert "Connection error. Unable to contact server."

I am using mitmproxy and http toolkit 'connect through adb'. The phone is rooted and I have su access. HTTPtoolkit gives all green checkmarks. 

I love all these tools and use the cert pinning for other apps, but Roblox is awfully stubborn! fwiw, when I do a tcpdump of playing Roblox on macos (instead of on android) I see the RakNet protocol for all the Roblox traffic (encrypted). 

Thanks for all wisdom! And thanks for all the great docs and open source code in the past!

**Comment by pimterry** [comment]
I've got a guide for cases that don't work out of the box that you might find helpful here: https://httptoolkit.com/blog/android-reverse-engineering/.

In general, the hard part is working out where the actual pinning is happening. It's a game of digging through logs, searching the source, and using Frida to monitor calls to every kind of exception throwing or relevant APIs that you can think of.

Something within Roblox must be trying to make a request that's then rejecting the connection. If you can find the code that's actually doing the check and failing, share it here and I'm happy to help write up a patch to work around it (and potentially integrate that here, if possible).

Is there anything shown in the ADB logs, or any errors listed in the Frida script output that might provide some clues? A stacktrace would be perfect, but any kinds of hints as to where exactly in the code this is being rejected would be helpful.

**Comment by pimterry** [comment]
I don't think that code posted above is relevant - that does throw `SSLPeerUnverifiedException` but only after a `InvocationTargetException` which I suspect is related to some lower level or parsing error on the certificate (which shouldn't happen in these cases) rather than certificate pinning.

The logs are very interesting though - that `TlsVerificationFail` definitely sounds like the issue, and like a unique error we should be able to follow.

Unfortunately though, searching the APK, that doesn't appear literally in the code - only inside the binary native file in `resources/lib/$ARCHITECTURE/libroblox.so`.

I suspect that means Roblox is shipping its own native TLS library, bundled inside that SO file. This could be for cross-platform compatibility & development benefits, since it means they can build once, compile natively, and then run the same code everywhere. It's inconvenient for us though, since this means that modifying this logic will require reverse engineering that, and that is definitely going to be challenging (and beyond what I'm familiar with I'm afraid - I'm more comfortable on the Java side here).

One more promising clue though is `resources/assets/cacert.pem`. That file in the APK contains a long list of certificates. I strongly suspect that's a limited set of CAs that the app trusts! If you modify that, then rebuild the APK and install it, that might well work for you.

**Comment by pimterry** [comment]
There's also `resources/assets/fingerprint.txt`, which could very likely be a certificate fingerprint used for pinning.

**Comment by antiworkink** [comment]
hey did you find a fix?

**Comment by Aze1337** [comment]
i believe the only way is to get good at reverse engineering mobile applications, to figure out where the ssl pinning happens, and from that attempt to replace it, remove it, or whatever that can be done, that needs some really good experience of reverse engineering, there is no one-time installation plugin that could fix that. its simply a security put in place so that the requests are not as easily accessible to you.
